snorkel/classification/loss.py

```diff
@@ -0,0 +1,122 @@
+from typing import Callable, List, Mapping, Optional, Sequence, Tuple, Union
+
+import torch
+import torch.nn.functional as F
+
+Outputs = Mapping[str, List[torch.FloatTensor]]
+
+
+def cross_entropy_soft(
+    input: torch.FloatTensor,
+    target: torch.FloatTensor,
+    weight: Optional[torch.FloatTensor] = None,
+    reduction: str = "mean",
+) -> torch.Tensor:
+    """Calculate cross-entropy loss when targets are probabilities (floats), not ints
+
+    PyTorch's F.cross_entropy() method requires integer labels; it does accept
+    probabilistic labels. We can, however, simulate such functionality with a for loop,
+    calculating the loss contributed by each class and accumulating the results.
+    Libraries such as keras do not require this workaround, as methods like
+    "categorical_crossentropy" accept float labels natively.
+
+    Note that the method signature is intentionally very similar to F.cross_entropy()
+    so that it can be used as a drop-in replacement when target labels are changed from
+    from a 1D tensor of ints to a 2D tensor of probabilities.
+
+    Parameters
+    ----------
+    input
+        A [num_points, num_classes] tensor of logits
+    target
+        A [num_points, num_classes] tensor of probabilistic target labels
+    weight
+        An optional [num_classes] array of weights to multiply the loss by per class
+    reduction
+        One of "none", "mean", "sum", indicating whether to return one loss per data
+        point, the mean loss, or the sum of losses
+
+    Returns
+    -------
+    torch.Tensor
+        The calculated loss
+
+    Raises
+    ------
+    ValueError
+        If an invalid reduction keyword is submitted
+    """
+    num_points, num_classes = input.shape
+    # Note that t.new_zeros, t.new_full put tensor on same device as t
+    cum_losses = input.new_zeros(num_points)
+    for y in range(num_classes):
```

Comment:
Why not just vectorize? a la https://github.com/tensorflow/tensorflow/blob/r1.14/tensorflow/python/keras/backend.py#L4149

Reply:
Is F.cross_ent not able to operate over axes?

Reply:
I wish. F.cross_entropy() has no axis arg; input has to be [num_points, num_classes] and target has to be [num_points]. It really is painful having a for loop in the loss calculation, but I think writing our own vectorized version not in cuda would still be slower.